org.lflang.tests/src/org/lflang/tests/TestBase.java

```diff
@@ -426,48 +456,163 @@ private void generateCode(LFTest test) {
+                        return pathName.getName().endsWith("Dockerfile");
+                    }
+                });
+                assert dockerFile.length == 1;
+                fedNameToDockerFile.put(srcGenFile.getName(), dockerFile[0].getAbsoluteFile().toPath());
+            }
+        }
+        return fedNameToDockerFile;
+    }
+
+    /**
+     * Return the content of the bash script used for testing docker option in federated execution. 
+     * @param fedNameToDockerFile A mapping of federateName -> absolutePathToDockerFile
+     * @param testNetworkName The name of the network used for testing the docker option. 
+     *                        See https://github.com/lf-lang/lingua-franca/wiki/Containerized-Execution#federated-execution for more details.
+     */
+    private String getDockerRunScript(Map<String, Path> fedNameToDockerFile, String testNetworkName) {
+        StringBuilder shCode = new StringBuilder();
+        shCode.append("#!/bin/bash\n");
+        int n = fedNameToDockerFile.size();
+        shCode.append("pids=\"\"\n");
+        shCode.append(String.format("docker run --rm --network=%s --name=rti rti:test -i 1 -n %d &\n", testNetworkName, n));
+        shCode.append("pids+=\"$!\"\nsleep 3\n");
+        for (String fedName : fedNameToDockerFile.keySet()) {
+            Path dockerFile = fedNameToDockerFile.get(fedName);
+            shCode.append(String.format("docker run --rm --network=%s %s:test -i 1 &\n", testNetworkName, fedName));
+            shCode.append("pids+=\" $!\"\n");
+        }
+        shCode.append("for p in $pids; do\n");
+        shCode.append("    if wait $p; then\n");
+        shCode.append("        :\n");
+        shCode.append("    else\n");
+        shCode.append("        exit 1\n");
+        shCode.append("    fi\n");
+        shCode.append("done\n");
+        return shCode.toString();
+    }
+
+    /**
+     * Return a list of ProcessBuilders used to test the docker option under non-federated execution.
+     * @param test The test to get the execution command for.
+     */
+    private List<ProcessBuilder> getNonfederatedDockerExecCommand(LFTest test) {
+        var srcGenPath = test.fileConfig.getSrcGenPath();
+        var dockerPath = srcGenPath.resolve(test.fileConfig.name + ".Dockerfile");
+        return Arrays.asList(new ProcessBuilder("docker", "build", "-t", "lingua_franca:test", "-f", dockerPath.toString(), srcGenPath.toString()), 
+                             new ProcessBuilder("docker", "run", "--rm", "lingua_franca:test"),
+                             new ProcessBuilder("docker", "image", "rm", "lingua_franca:test"));
+    }
+
+    /**
+     * Return a list of ProcessBuilders used to test the docker option under federated execution.
+     * @param test The test to get the execution command for.
+     */
+    private List<ProcessBuilder> getFederatedDockerExecCommand(LFTest test) {
+        var rtiPath = test.fileConfig.getSrcGenBasePath().resolve("RTI");
```

Comment:
At some point, there should be a test that `docker` actually exists, and print an informative message to the user if not. We have similar checks for things like `gcc` somewhere. Remember that a user can run these tests locally.

Reply:
one way to do it is to run `which docker` and a non-zero exit code indicates that docker is missing.

Reply:
We have an [`LFCommand`](https://github.com/lf-lang/lingua-franca/blob/ce52695f2a07296cd63ae1a71cc05b45ad377b57/org.lflang/src/org/lflang/util/LFCommand.java#L50) class that already does this. There are some nuances, such as using `where` instead of `which` on Windows.